### PR TITLE
[SPARK-28002][SQL] Support WITH clause column aliases

### PR DESCRIPTION
**Pull request by peter-toth**
## What changes were proposed in this pull request?

This PR adds support of column aliasing in a CTE so this query becomes valid:
```
WITH t(x) AS (SELECT 1)
SELECT * FROM t WHERE x = 1
```
## How was this patch tested?

Added new UTs.

### CODE REVIEW

**Comment by peter-toth** [comment]
@dongjoon-hyun, @gatorsmile this is another feature that PostgreSQL does support.

**Comment by liancheng** [comment]
ok to test

**Comment by liancheng** [comment]
LGTM pending Jenkins, thanks!

**Comment by liancheng** [comment]
In fact, I was thinking, instead of adding test cases in the `SQLQuerySuite`, maybe it's better to add test cases in the [`PlanParserSuite`][1] and [`AnalysisSuite`][2].

The changes in this PR are purely about the parsing phase and running full-blown SQL queries for parsing changes is slow and unnecessary. The Spark PR builder already takes a long time to finish, would be better to cut the cost whenever possible.

[1]: https://github.com/apache/spark/blob/1217996f1574f758d8cccc1c4e3846452d24b35b/sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/PlanParserSuite.scala
[2]: https://github.com/apache/spark/blob/1217996f1574f758d8cccc1c4e3846452d24b35b/sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisSuite.scala

**Comment by peter-toth** [comment]
Thanks @liancheng for the review. I've moved analysis error test cases to `AnalysisSuite`. But I have left the positive test case in `SQLQuerySuite` as I think the value of `x` column should be tested.

**Comment by hvanhovell** [comment]
@peter-toth can you still add a test to the plan parser suite?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 1]
```diff
@@ -161,9 +161,14 @@ class AstBuilder(conf: SQLConf) extends SqlBaseBaseVisitor[AnyRef] with Logging
    *
    * This is only used for Common Table Expressions.
    */
-  override def visitNamedQuery(ctx: NamedQueryContext): SubqueryAlias = withOrigin(ctx) {
-    SubqueryAlias(ctx.name.getText, plan(ctx.query))
-  }
+  override def visitNamedQuery(ctx: NamedQueryContext): SubqueryAlias =
+    withOrigin(ctx) {
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: nit. Shall we keep the original form like the following? That reduces both the diff of patch and the depth of indentation of the function body.
```scala
override def visitNamedQuery(ctx: NamedQueryContext): SubqueryAlias = withOrigin(ctx) {
```

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisSuite.scala [hunk 2]
```diff
@@ -633,4 +634,16 @@ class AnalysisSuite extends AnalysisTest with Matchers {
...
+    assertAnalysisError(parsePlan("WITH t(x) AS (SELECT 1) SELECT * FROM t WHERE y = 1"),
+      Seq("cannot resolve '`y`' given input columns: [x]"))
+  }
+
+  test("SPARK-28002: CTE with non-matching column alias") {
+    assertAnalysisError(parsePlan("WITH t(x, y) AS (SELECT 1) SELECT * FROM t WHERE x = 1"),
+      Seq("Number of column aliases does not match number of columns. Number of column aliases: " +
+        "2; number of columns: 1."))
+  }
+
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisSuite.scala]: nit. remove empty line.

**Comment by dongjoon-hyun** [comment]
Thank you for ping me, @peter-toth . +1 for the above test case comments. And, this PR also looks good to me.

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 3]
```diff
@@ -545,6 +545,16 @@ class SQLQuerySuite extends QueryTest with SharedSQLContext {
...
 
+  test("SPARK-28002: CTE with column alias") {
+    checkAnswer(
+      sql("""
+        |WITH t(x) AS (SELECT 1)
+        |SELECT * FROM t WHERE x = 1
+        """.stripMargin),
+      Row(1) :: Nil
+    )
+  }
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: Oh, please move this to `cite.sql`. That is a perfect place for this.
cc @gatorsmile

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala (hunk 1 above)
Comment (peter-toth) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: thanks, reverted to the original form

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisSuite.scala (hunk 2 above)
Comment (peter-toth) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisSuite.scala]: ok, done

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala (hunk 3 above)
Comment (peter-toth) [reply sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: ok, moved

**Comment by peter-toth** [comment]
> @peter-toth can you still add a test to the plan parser suite?

@hvanhovell sure, I added a test to `PlanParserSuite`

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisSuite.scala [hunk 4]
```diff
@@ -633,4 +634,15 @@ class AnalysisSuite extends AnalysisTest with Matchers {
     val res = ViewAnalyzer.execute(view)
     comparePlans(res, expected)
   }
+
+  test("SPARK-28002: CTE with non-existing column alias") {
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisSuite.scala]: For new feature test case, we don't use SPARK issue id. Let's remove this `SPARK-28002: `.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisSuite.scala [hunk 5]
```diff
@@ -633,4 +634,15 @@ class AnalysisSuite extends AnalysisTest with Matchers {
...
+    assertAnalysisError(parsePlan("WITH t(x) AS (SELECT 1) SELECT * FROM t WHERE y = 1"),
+      Seq("cannot resolve '`y`' given input columns: [x]"))
+  }
+
+  test("SPARK-28002: CTE with non-matching column alias") {
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisSuite.scala]: ditto.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/PlanParserSuite.scala [hunk 6]
```diff
@@ -801,4 +807,10 @@ class PlanParserSuite extends AnalysisTest {
     }.getMessage
     assert(m2.contains("mismatched input 'FROM' expecting"))
   }
+
+  test("SPARK-28002: CTE with column alias") {
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/PlanParserSuite.scala]: ditto.

File: sql/core/src/test/resources/sql-tests/inputs/cte.sql [hunk 7]
```diff
@@ -24,6 +24,10 @@ SELECT t1.id AS c1,
 FROM   CTE1 t1
        CROSS JOIN CTE1 t2;
 
+-- SPARK-28002: CTE with column alias
```
Comment (dongjoon-hyun) [inline sql/core/src/test/resources/sql-tests/inputs/cte.sql]: Here, too. We can remove this comment.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/PlanParserSuite.scala [hunk 8]
```diff
@@ -81,18 +81,31 @@ class PlanParserSuite extends AnalysisTest {
...
+      case (name, (cte, columnAliases)) =>
+        val subquery = if (columnAliases.isEmpty) {
+          cte
+        } else {
+          UnresolvedSubqueryColumnAliases(columnAliases, cte)
+        }
+        name -> SubqueryAlias(name, subquery)
+    }
+    With(plan, ctes)
+  }
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/PlanParserSuite.scala]: This test suite is updated a few minute ago. Could you rebase once more?
There will be another `cte` function in this test suite.

Comment (dongjoon-hyun) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/PlanParserSuite.scala]: Oh, this PR is already updated to the master. Then, please move this function to the below of the existing `cte` function. We had better gather the similar functions. If possible, please consolidate them into one.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisSuite.scala (hunk 4 above)
Comment (peter-toth) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisSuite.scala]: thanks, fixed

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/PlanParserSuite.scala (hunk 8 above)
Comment (peter-toth) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/PlanParserSuite.scala]: Fixed, thanks.

**Comment by peter-toth** [comment]
retest this please

File: sql/core/src/test/resources/sql-tests/inputs/cte.sql [hunk 9]
```diff
@@ -24,6 +24,10 @@ SELECT t1.id AS c1,
 FROM   CTE1 t1
        CROSS JOIN CTE1 t2;
 
+-- CTE with column alias
+WITH t(x) AS (SELECT 1)
```
Comment (gatorsmile) [inline sql/core/src/test/resources/sql-tests/inputs/cte.sql]: Could you add more test cases? For example, can a with clause in a subquery shadow a with clause in an enclosing query with the same name? Another example, use with clauses in a subquery expression?

Comment (peter-toth) [reply sql/core/src/test/resources/sql-tests/inputs/cte.sql]: I wanted to focus on column aliases in this PR, but here is another open here which focuses on nested `WITH` clauses: https://github.com/apache/spark/pull/24831

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/PlanParserSuite.scala [hunk 10]
```diff
@@ -42,10 +42,15 @@ class PlanParserSuite extends AnalysisTest {
   private def intercept(sqlCommand: String, messages: String*): Unit =
     interceptParseException(parsePlan)(sqlCommand, messages: _*)
 
-  private def cte(plan: LogicalPlan, namedPlans: (String, LogicalPlan)*): With = {
+  private def cte(plan: LogicalPlan, namedPlans: (String, (LogicalPlan, Seq[String]))*): With = {
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/PlanParserSuite.scala]: Thank you for consolidating both `cte` functions into one!

File: sql/core/src/test/resources/sql-tests/inputs/cte.sql (hunk 9 above)
Comment (dongjoon-hyun) [reply sql/core/src/test/resources/sql-tests/inputs/cte.sql]: Yep. For the nested `WITH`, #24831 would be a better place to add them.
After merging this, let's proceed as @gatorsmile suggested, @peter-toth .

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM.

@gatorsmile . Could you do the final review and sign-off this? For the nested `WITH`, I hope we can do that in his another PRs.

- #24831 is adding nested WITH.
- #24860 is adding WITH test cases.

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by dongjoon-hyun** [comment]
Merged to master. 

Thank you so much, @peter-toth , @liancheng, @hvanhovell , @gatorsmile !

@peter-toth . Please proceed to #24831 and #24860 .

**Comment by peter-toth** [comment]
Thanks @dongjoon-hyun @liancheng @gatorsmile @hvanhovell for the review.

I have prepared https://github.com/apache/spark/pull/24831 for review.
Unfortunately I can't work on https://github.com/apache/spark/pull/24860 next week but you can expect updates on it after that.

**Comment by dongjoon-hyun** [comment]
Thanks. No problem, @peter-toth !

File: sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4 [hunk 11]
```diff
@@ -319,7 +319,7 @@ ctes
     ;
 
 namedQuery
-    : name=identifier AS? '(' query ')'
+    : name=identifier (columnAliases=identifierList)? AS? '(' query ')'
```
Comment (gatorsmile) [inline sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4]: Duplicate names within a single CTE definition are not allowed.

File: sql/core/src/test/resources/sql-tests/results/cte.sql.out [hunk 12]
```diff
@@ -89,16 +89,25 @@ struct<c1:int,c2:int>
 
 
 -- !query 7
-DROP VIEW IF EXISTS t
+WITH t(x) AS (SELECT 1)
+SELECT * FROM t WHERE x = 1
```
Comment (gatorsmile) [inline sql/core/src/test/resources/sql-tests/results/cte.sql.out]: Almost all the test cases are using one column in the CTE definition. Can you try your best to improve the current test coverage in this new syntax?

**Comment by gatorsmile** [comment]
Sorry for my late response. @peter-toth @dongjoon-hyun Could you submit a follow-up PR to improve it?

**Comment by peter-toth** [comment]
@gatorsmile @dongjoon-hyun I opened https://github.com/apache/spark/pull/24949 to add some new test cases. Please let me know if you want more cases.

Please note that I'm also working on https://github.com/apache/spark/pull/24860 and it will add many new tests that cover WITH column aliases.

**Pull request closed without merging**